lib/signex/algorithm.ex

```diff
@@ -0,0 +1,74 @@
+defmodule SignEx.Algorithm do
+  @moduledoc """
+    A `SignEx.Algorithm` struct contains all the information needed to deal with signature encryption and digest
+  """
+
+  @enforce_keys [:digest]
+  defstruct [:encryption, :digest]
+
+  defimpl String.Chars, for: SignEx.Algorithm do
+    def to_string(alg), do: "#{alg.encryption}-#{alg.digest}"
+  end
+
+  @available [
+        {:ec, :sha256},
+        {:ec, :sha384},
+        {:ec, :sha512},
+        {:rsa, :sha256},
+        {:rsa, :sha384},
+        {:rsa, :sha512},
+      ]
+
+  @allowed_strings(
+    @available
+    |> Enum.map(&(to_string(elem(&1,0)) <> "-" <> to_string(elem(&1, 1))))
+  )
+  @doc """
+  List of available algorithms where each element is SignEx.Algorithm struct
+  """
+  defmacro available do
+    quote do
+      unquote(@available)
+      |> Enum.map(fn({encryption, digest}) ->
+        %SignEx.Algorithm{encryption: encryption, digest: digest}
+      end)
+    end
+  end
+
+  @doc """
+  List of allowed algorithm string where each element is in format 'encryption-digest'
+  """
+  defmacro allowed_strings, do: @allowed_strings
```

Comment:
Does this work? I would have expected it to need to be `def` not `defmacro` as a macro needs to return an ast